# Aim

The purpose of this document is to write down a list of publicly available datasets that may be suitable for teaching. 


# Non-biological

+ `gapminder`



# Gene expression 

+ Take this sample data and add on 10 genes from NanoString https://www.ncbi.nlm.nih.gov/geo/geo2r/?acc=GSE43251&platform=GPL16444. This data has a pretty good experimental design. https://rnajournal.cshlp.org/content/19/1/51/suppl/DC1 supplementary table 4 has the sample labelling. 

+ This sample data is pretty dirty with replicates. https://www.ncbi.nlm.nih.gov/geo/geo2r/?acc=GSE55055. Not great, but possible. 

+ This data is pretty interesting. https://www.ncbi.nlm.nih.gov/geo/geo2r/?acc=GSE59595&platform=GPL8179. Breast cancer patients from China and Italy with differnt subtypes. Maybe we add 10 genes to make sure that there are some additional numerical columns to play with. 

# Other biological data

+ This example used Gene Ontology as the example dataset. https://hbctraining.github.io/Training-modules/Tidyverse_ggplot2/lessons/intro_tidyverse.html



# Other resources

+ https://girke.bioinformatics.ucr.edu/GEN242/mydoc_Rbasics_12.html has very comprehensive materials on different aspects of bioinformatics. 

+ https://monashbioinformaticsplatform.github.io/r-more/topics/tidyverse.html has some basic cleaning and visualisation for bioinformatics task. https://github.com/monashbioinformaticsplatform. https://www.monash.edu/researchinfrastructure/bioinformatics/training. 

+ https://www.bioinformatics.babraham.ac.uk/training/Tidyverse/tidyverse_final_exercise.html#expression: tidyverse and mythylation data. 

+ From Garth, http://garthtarr.github.io/visR/labs/01/index.html. Which was built on http://varianceexplained.org/r/tidy-genomics/. 