<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Kevin Wang and Garth Tarr" />

<meta name="date" content="2019-05-11" />

<title>session3</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script defer src="https://pro.fontawesome.com/releases/v5.0.13/js/all.js" integrity="sha384-d84LGg2pm9KhR4mCAs3N29GQ4OYNy+K+FBHX8WhimHpPm86c839++MDABegrZ3gn" crossorigin="anonymous"></script>

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<script src="clipboard.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    background-color: #ffffff;
    color: #a0a0a0;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #a0a0a0;  padding-left: 4px; }
div.sourceCode
  { color: #1f1c1b; background-color: #ffffff; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span. { color: #1f1c1b; } /* Normal */
code span.al { color: #bf0303; background-color: #f7e6e6; font-weight: bold; } /* Alert */
code span.an { color: #ca60ca; } /* Annotation */
code span.at { color: #0057ae; } /* Attribute */
code span.bn { color: #b08000; } /* BaseN */
code span.bu { color: #644a9b; font-weight: bold; } /* BuiltIn */
code span.cf { color: #1f1c1b; font-weight: bold; } /* ControlFlow */
code span.ch { color: #924c9d; } /* Char */
code span.cn { color: #aa5500; } /* Constant */
code span.co { color: #898887; } /* Comment */
code span.cv { color: #0095ff; } /* CommentVar */
code span.do { color: #607880; } /* Documentation */
code span.dt { color: #0057ae; } /* DataType */
code span.dv { color: #b08000; } /* DecVal */
code span.er { color: #bf0303; text-decoration: underline; } /* Error */
code span.ex { color: #0095ff; font-weight: bold; } /* Extension */
code span.fl { color: #b08000; } /* Float */
code span.fu { color: #644a9b; } /* Function */
code span.im { color: #ff5500; } /* Import */
code span.in { color: #b08000; } /* Information */
code span.kw { color: #1f1c1b; font-weight: bold; } /* Keyword */
code span.op { color: #1f1c1b; } /* Operator */
code span.ot { color: #006e28; } /* Other */
code span.pp { color: #006e28; } /* Preprocessor */
code span.re { color: #0057ae; background-color: #e0e9f8; } /* RegionMarker */
code span.sc { color: #3daee9; } /* SpecialChar */
code span.ss { color: #ff5500; } /* SpecialString */
code span.st { color: #bf0303; } /* String */
code span.va { color: #0057ae; } /* Variable */
code span.vs { color: #bf0303; } /* VerbatimString */
code span.wa { color: #bf0303; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">BioinfoSummer - Tidyverse</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="session1.html">Mon: session 1</a>
</li>
<li>
  <a href="session2.html">Tue: session 2</a>
</li>
<li>
  <a href="session3.html">Tue: session 3</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">session3</h1>
<h4 class="author">Kevin Wang and Garth Tarr</h4>
<h4 class="date">05/11/2019</h4>

</div>


<div id="learning-outcomes-of-this-session" class="section level1">
<h1>Learning outcomes of this session</h1>
<ul>
<li>More plotting with <code>ggplot2</code></li>
<li><code>for</code> loops and <code>purrr::map</code> functions</li>
<li>Modelling with <code>broom</code></li>
</ul>
</div>
<div id="loading-packages" class="section level1">
<h1>Loading packages</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(tidyverse)</span></code></pre></div>
</div>
<div id="loading-cleaned-sample-data" class="section level1">
<h1>Loading cleaned sample data</h1>
<p>This session3 data is almost identical to the <code>wide_sample_ge_data.csv</code> file that we just saved. Except that we actually used all the genes in the database.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>session3_data =<span class="st"> </span><span class="kw">read_csv</span>(<span class="dt">file =</span> <span class="st">&quot;data/session3_data.csv&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span>time, <span class="op">-</span>dose)</span></code></pre></div>
<pre><code>## Parsed with column specification:
## cols(
##   .default = col_double(),
##   strain = col_character(),
##   sample = col_character()
## )</code></pre>
<pre><code>## See spec(...) for full column specifications.</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>clean_data =<span class="st"> </span>session3_data <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">mutate_if</span>(is.numeric, <span class="dt">.funs =</span> log)</span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a>clean_data</span></code></pre></div>
<pre><code>## # A tibble: 154 x 56
##    strain sample  Ahrr Aldh3a1  Bbs2 Ccbl1 Cxcr7 Col18a1 Cyb5a Cyp1a1
##    &lt;chr&gt;  &lt;chr&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
##  1 HW     HW1    2.57     1.65  3.52  5.47  3.16    7.43 10.1    2.90
##  2 HW     HW2    0.352    2.14  3.82  5.49  3.30    7.56 10.2    4.22
##  3 HW     HW3    0.254    2.05  3.47  5.17  2.74    6.82 10.1    3.47
##  4 HW     HW4    2.30     5.73  4.24  5.62  4.22    7.77 10.4    9.20
##  5 HW     HW5    2.02     4.18  3.97  5.76  3.53    7.43 10.2    8.84
##  6 HW     HW6    1.70     4.40  3.87  5.50  3.70    7.78 10.00   8.88
##  7 HW     WW45   2.90     7.98  4.13  6.34  3.62    7.19  9.90  10.7 
##  8 HW     WW46   3.11     7.01  4.30  6.26  3.73    7.05  9.85  10.3 
##  9 HW     WW47   3.36     7.73  3.97  6.48  4.31    7.49  9.83  10.7 
## 10 HW     WW48   2.04     7.96  3.74  6.26  3.85    6.90  9.67  10.3 
## # … with 144 more rows, and 46 more variables: Cyp1a2 &lt;dbl&gt;, Cyp1b1 &lt;dbl&gt;,
## #   Cyp2a1 &lt;dbl&gt;, Cyp2s1 &lt;dbl&gt;, Eef1a1 &lt;dbl&gt;, Eml4 &lt;dbl&gt;, Exoc3 &lt;dbl&gt;,
## #   Fdft1 &lt;dbl&gt;, Fmo1 &lt;dbl&gt;, Gapdh &lt;dbl&gt;, Gfer &lt;dbl&gt;, Ghr &lt;dbl&gt;,
## #   Glud1 &lt;dbl&gt;, Hagh &lt;dbl&gt;, Hprt1 &lt;dbl&gt;, Hsd17b12 &lt;dbl&gt;, Ivns1abp &lt;dbl&gt;,
## #   Lasp1 &lt;dbl&gt;, Uvrag &lt;dbl&gt;, Inmt &lt;dbl&gt;, Pbld &lt;dbl&gt;, Mug1 &lt;dbl&gt;,
## #   Neu1 &lt;dbl&gt;, Nfe2l2 &lt;dbl&gt;, Nqo1 &lt;dbl&gt;, Nrbp2 &lt;dbl&gt;, Nudt4 &lt;dbl&gt;,
## #   Pde2a &lt;dbl&gt;, Perp &lt;dbl&gt;, Pgk1 &lt;dbl&gt;, Pmm1 &lt;dbl&gt;, Ppia &lt;dbl&gt;,
## #   Psmb4 &lt;dbl&gt;, Pomp &lt;dbl&gt;, Derl1 &lt;dbl&gt;, Ern1 &lt;dbl&gt;, Rpl13a &lt;dbl&gt;,
## #   Scn4b &lt;dbl&gt;, Sdc1 &lt;dbl&gt;, Sdha &lt;dbl&gt;, Slco1a1 &lt;dbl&gt;, Srxn1 &lt;dbl&gt;,
## #   Tiparp &lt;dbl&gt;, Tpm1 &lt;dbl&gt;, Ugt1a1 &lt;dbl&gt;, Zfp422 &lt;dbl&gt;</code></pre>
</div>
<div id="t-test" class="section level1">
<h1>t-test</h1>
<p>A typical task in bioinformatics is to perform differential gene expression analysis. That is, we want to identify genes that has a different expression patterns between two conditions. While there are many ways that this can be done, we will opt for the simplest approach of using a t-test.</p>
<p>In <code>R</code>, t-test can be performed using the <code>t.test</code> function. Below, we will perform a t-test on the Ahrrr gene between the two strains of mice.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">t.test</span>(Ahrr <span class="op">~</span><span class="st"> </span>strain, <span class="dt">data =</span> clean_data)</span></code></pre></div>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  Ahrr by strain
## t = -0.75737, df = 148.97, p-value = 0.45
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -0.5485616  0.2445692
## sample estimates:
## mean in group HW mean in group LE 
##         2.998168         3.150164</code></pre>
<p>However, the output of the <code>t.test</code> function can be a bit hard to read or manipulate into a data.frame. This is where the <code>tidy()</code> function from the <code>broom</code> package can help.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="kw">t.test</span>(Ahrr <span class="op">~</span><span class="st"> </span>strain, <span class="dt">data =</span> clean_data) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="st">  </span>broom<span class="op">::</span><span class="kw">tidy</span>()</span></code></pre></div>
<pre><code>## # A tibble: 1 x 10
##   estimate estimate1 estimate2 statistic p.value parameter conf.low
##      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1   -0.152      3.00      3.15    -0.757   0.450      149.   -0.549
## # … with 3 more variables: conf.high &lt;dbl&gt;, method &lt;chr&gt;,
## #   alternative &lt;chr&gt;</code></pre>
<p>Similarly, we can do the same for the Cxcr7 gene.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw">t.test</span>(Cxcr7 <span class="op">~</span><span class="st"> </span>strain, <span class="dt">data =</span> clean_data) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="st">  </span>broom<span class="op">::</span><span class="kw">tidy</span>()</span></code></pre></div>
<pre><code>## # A tibble: 1 x 10
##   estimate estimate1 estimate2 statistic p.value parameter conf.low
##      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1    -1.19      3.57      4.76     -6.74 1.34e-9      92.9    -1.54
## # … with 3 more variables: conf.high &lt;dbl&gt;, method &lt;chr&gt;,
## #   alternative &lt;chr&gt;</code></pre>
</div>
<div id="doing-multiple-t-tests" class="section level1">
<h1>Doing multiple t-tests</h1>
<p>In the <code>clean_data</code>, we have 54 genes in total. If we want to perform t-test on each gene, should we type out 54 lines of code?</p>
<p>The answer is obviously no. Because such practice is very time inefficient and not generalisable if we receive another dataset in the future.</p>
<p>Now, let’s think about this very difficult task (in a more abstract level). First, we simply run the <code>t.test</code> and <code>broom::tidy</code> function on gene expression values for a selected gene. Then, we do exactly the same operation for another 53 genes. So the whole task is repetitive and we really are apply the following function to 54 datasets (each dataset only contains one selected gene).</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>tidy_test =<span class="st"> </span><span class="cf">function</span>(this_data){</span>
<span id="cb13-2"><a href="#cb13-2"></a>  <span class="kw">t.test</span>(gene_expression <span class="op">~</span><span class="st"> </span>strain, <span class="dt">data =</span> this_data) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="st">    </span>broom<span class="op">::</span><span class="kw">tidy</span>()</span>
<span id="cb13-4"><a href="#cb13-4"></a>}</span></code></pre></div>
<p>So now, our very difficult task of performing 54 t-tests is simplified to spliting <code>clean_data</code> into 54 datasets, one for each gene.</p>
<p>While there are many ways this can be done in <code>R</code>, we will use a <code>tidyverse</code> framework that has the advantage of readable codes and generalisability.</p>
<p>First, let’s pivot our data into a long format.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>gene_long =<span class="st"> </span>clean_data <span class="op">%&gt;%</span></span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">pivot_longer</span>(</span>
<span id="cb14-3"><a href="#cb14-3"></a>    <span class="dt">cols =</span> <span class="op">-</span><span class="kw">c</span>(<span class="st">&quot;strain&quot;</span>, <span class="st">&quot;sample&quot;</span>),</span>
<span id="cb14-4"><a href="#cb14-4"></a>    <span class="dt">names_to =</span> <span class="st">&quot;gene_symbols&quot;</span>,</span>
<span id="cb14-5"><a href="#cb14-5"></a>    <span class="dt">values_to =</span> <span class="st">&quot;gene_expression&quot;</span>)</span>
<span id="cb14-6"><a href="#cb14-6"></a></span>
<span id="cb14-7"><a href="#cb14-7"></a>gene_long</span></code></pre></div>
<pre><code>## # A tibble: 8,316 x 4
##    strain sample gene_symbols gene_expression
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;                  &lt;dbl&gt;
##  1 HW     HW1    Ahrr                    2.57
##  2 HW     HW1    Aldh3a1                 1.65
##  3 HW     HW1    Bbs2                    3.52
##  4 HW     HW1    Ccbl1                   5.47
##  5 HW     HW1    Cxcr7                   3.16
##  6 HW     HW1    Col18a1                 7.43
##  7 HW     HW1    Cyb5a                  10.1 
##  8 HW     HW1    Cyp1a1                  2.90
##  9 HW     HW1    Cyp1a2                 10.6 
## 10 HW     HW1    Cyp1b1                  1.87
## # … with 8,306 more rows</code></pre>
<p>Then we will use the <code>group_by</code> function and then apply the <code>nest</code> function from the <code>tidyr</code> package.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>gene_nest =<span class="st"> </span>gene_long <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb16-2"><a href="#cb16-2"></a><span class="st">  </span><span class="kw">group_by</span>(gene_symbols) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">nest</span>()</span>
<span id="cb16-4"><a href="#cb16-4"></a></span>
<span id="cb16-5"><a href="#cb16-5"></a>gene_nest</span></code></pre></div>
<pre><code>## # A tibble: 54 x 2
## # Groups:   gene_symbols [54]
##    gene_symbols           data
##    &lt;chr&gt;        &lt;list&lt;df[,3]&gt;&gt;
##  1 Ahrr              [154 × 3]
##  2 Aldh3a1           [154 × 3]
##  3 Bbs2              [154 × 3]
##  4 Ccbl1             [154 × 3]
##  5 Cxcr7             [154 × 3]
##  6 Col18a1           [154 × 3]
##  7 Cyb5a             [154 × 3]
##  8 Cyp1a1            [154 × 3]
##  9 Cyp1a2            [154 × 3]
## 10 Cyp1b1            [154 × 3]
## # … with 44 more rows</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>gene_nest<span class="op">$</span>data[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</span></code></pre></div>
<pre><code>## &lt;list_of&lt;
##   tbl_df&lt;
##     strain         : character
##     sample         : character
##     gene_expression: double
##   &gt;
## &gt;[2]&gt;
## [[1]]
## # A tibble: 154 x 3
##    strain sample gene_expression
##    &lt;chr&gt;  &lt;chr&gt;            &lt;dbl&gt;
##  1 HW     HW1              2.57 
##  2 HW     HW2              0.352
##  3 HW     HW3              0.254
##  4 HW     HW4              2.30 
##  5 HW     HW5              2.02 
##  6 HW     HW6              1.70 
##  7 HW     WW45             2.90 
##  8 HW     WW46             3.11 
##  9 HW     WW47             3.36 
## 10 HW     WW48             2.04 
## # … with 144 more rows
## 
## [[2]]
## # A tibble: 154 x 3
##    strain sample gene_expression
##    &lt;chr&gt;  &lt;chr&gt;            &lt;dbl&gt;
##  1 HW     HW1               1.65
##  2 HW     HW2               2.14
##  3 HW     HW3               2.05
##  4 HW     HW4               5.73
##  5 HW     HW5               4.18
##  6 HW     HW6               4.40
##  7 HW     WW45              7.98
##  8 HW     WW46              7.01
##  9 HW     WW47              7.73
## 10 HW     WW48              7.96
## # … with 144 more rows</code></pre>
<p>So, a lot of things happened here. Let’s explain.</p>
<p>You might remember the <code>group_by</code> function when we were trying to use the <code>summarise</code> function to calculate summary statistics. This is exactly what we want to achieve here with <code>gene_nest</code>. Each gene used to have 154 observations (number rows of <code>clean_data</code>), but in <code>gene_nest</code>, each gene only has “one observation”, which corresponds to a row in the <code>data</code> column. <strong>Each row of the <code>data</code> column is a <code>data.frame</code> of size 154 rows and 3 columns</strong>, indicated by the <code>[154 x 3]</code>. In each of the <code>[154 x 3]</code> data.frame, you will see three columns (all the columns that we didn’t put into <code>group_by</code>).</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>gene_test =<span class="st"> </span>gene_nest <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb20-2"><a href="#cb20-2"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(</span>
<span id="cb20-3"><a href="#cb20-3"></a>    <span class="dt">tidy_result =</span> purrr<span class="op">::</span><span class="kw">map</span>(</span>
<span id="cb20-4"><a href="#cb20-4"></a>      <span class="dt">.x =</span> data,</span>
<span id="cb20-5"><a href="#cb20-5"></a>      <span class="dt">.f =</span> tidy_test))</span>
<span id="cb20-6"><a href="#cb20-6"></a></span>
<span id="cb20-7"><a href="#cb20-7"></a>gene_model_long =<span class="st"> </span>gene_test <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb20-8"><a href="#cb20-8"></a><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">unnest</span>(tidy_result)</span>
<span id="cb20-9"><a href="#cb20-9"></a></span>
<span id="cb20-10"><a href="#cb20-10"></a></span>
<span id="cb20-11"><a href="#cb20-11"></a>gene_model_long <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb20-12"><a href="#cb20-12"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> p.value)) <span class="op">+</span></span>
<span id="cb20-13"><a href="#cb20-13"></a><span class="st">  </span><span class="kw">geom_histogram</span>()</span></code></pre></div>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="session3_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>gene_model_long <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb22-2"><a href="#cb22-2"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(p.value <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.0001</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb22-3"><a href="#cb22-3"></a><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">unnest</span>(data) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb22-4"><a href="#cb22-4"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(</span>
<span id="cb22-5"><a href="#cb22-5"></a>    <span class="dt">label =</span> <span class="kw">paste0</span>(gene_symbols, </span>
<span id="cb22-6"><a href="#cb22-6"></a>                   <span class="st">&quot;, p-value = &quot;</span>, </span>
<span id="cb22-7"><a href="#cb22-7"></a>                   <span class="kw">signif</span>(p.value, <span class="dv">2</span>))) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb22-8"><a href="#cb22-8"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> strain, </span>
<span id="cb22-9"><a href="#cb22-9"></a>             <span class="dt">y =</span> gene_expression)) <span class="op">+</span></span>
<span id="cb22-10"><a href="#cb22-10"></a><span class="st">  </span><span class="kw">geom_boxplot</span>() <span class="op">+</span></span>
<span id="cb22-11"><a href="#cb22-11"></a><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>label, <span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>)</span></code></pre></div>
<p><img src="session3_files/figure-html/unnamed-chunk-9-2.png" width="672" /></p>
</div>
<div id="regression-analysis-extension" class="section level1">
<h1>Regression analysis (extension)</h1>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>hw_data =<span class="st"> </span>clean_data <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb23-2"><a href="#cb23-2"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(strain <span class="op">==</span><span class="st"> &quot;HW&quot;</span>)</span>
<span id="cb23-3"><a href="#cb23-3"></a></span>
<span id="cb23-4"><a href="#cb23-4"></a>le_data =<span class="st"> </span>clean_data <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb23-5"><a href="#cb23-5"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(strain <span class="op">==</span><span class="st"> &quot;LE&quot;</span>)</span>
<span id="cb23-6"><a href="#cb23-6"></a></span>
<span id="cb23-7"><a href="#cb23-7"></a></span>
<span id="cb23-8"><a href="#cb23-8"></a>clean_data <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb23-9"><a href="#cb23-9"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> Ahrr,</span>
<span id="cb23-10"><a href="#cb23-10"></a>             <span class="dt">y =</span> Cxcr7,</span>
<span id="cb23-11"><a href="#cb23-11"></a>             <span class="dt">colour =</span> strain)) <span class="op">+</span></span>
<span id="cb23-12"><a href="#cb23-12"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb23-13"><a href="#cb23-13"></a><span class="st">  </span><span class="kw">geom_smooth</span>(<span class="dt">method =</span> <span class="st">&quot;lm&quot;</span>)</span></code></pre></div>
<p><img src="session3_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="kw">lm</span>(Cxcr7 <span class="op">~</span><span class="st"> </span>Ahrr, <span class="dt">data =</span> clean_data)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = Cxcr7 ~ Ahrr, data = clean_data)
## 
## Coefficients:
## (Intercept)         Ahrr  
##      2.2446       0.6054</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="kw">lm</span>(Cxcr7 <span class="op">~</span><span class="st"> </span>Ahrr, <span class="dt">data =</span> hw_data)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = Cxcr7 ~ Ahrr, data = hw_data)
## 
## Coefficients:
## (Intercept)         Ahrr  
##      2.5755       0.3311</code></pre>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="kw">lm</span>(Cxcr7 <span class="op">~</span><span class="st"> </span>Ahrr, <span class="dt">data =</span> le_data)</span></code></pre></div>
<pre><code>## 
## Call:
## lm(formula = Cxcr7 ~ Ahrr, data = le_data)
## 
## Coefficients:
## (Intercept)         Ahrr  
##      1.8324       0.9286</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a>lm_Cxcr7_Ahrr =<span class="st"> </span><span class="cf">function</span>(this_data){</span>
<span id="cb30-2"><a href="#cb30-2"></a>  <span class="kw">lm</span>(Cxcr7 <span class="op">~</span><span class="st"> </span>Ahrr, <span class="dt">data =</span> this_data)</span>
<span id="cb30-3"><a href="#cb30-3"></a>}</span>
<span id="cb30-4"><a href="#cb30-4"></a></span>
<span id="cb30-5"><a href="#cb30-5"></a>list_data =<span class="st"> </span><span class="kw">list</span>(</span>
<span id="cb30-6"><a href="#cb30-6"></a>  <span class="dt">clean_data =</span> clean_data, </span>
<span id="cb30-7"><a href="#cb30-7"></a>  <span class="dt">hw_data =</span> hw_data,</span>
<span id="cb30-8"><a href="#cb30-8"></a>  <span class="dt">le_data =</span> le_data)</span>
<span id="cb30-9"><a href="#cb30-9"></a></span>
<span id="cb30-10"><a href="#cb30-10"></a>purrr<span class="op">::</span><span class="kw">map</span>(<span class="dt">.x =</span> list_data, </span>
<span id="cb30-11"><a href="#cb30-11"></a>           <span class="dt">.f =</span> lm_Cxcr7_Ahrr)</span></code></pre></div>
<pre><code>## $clean_data
## 
## Call:
## lm(formula = Cxcr7 ~ Ahrr, data = this_data)
## 
## Coefficients:
## (Intercept)         Ahrr  
##      2.2446       0.6054  
## 
## 
## $hw_data
## 
## Call:
## lm(formula = Cxcr7 ~ Ahrr, data = this_data)
## 
## Coefficients:
## (Intercept)         Ahrr  
##      2.5755       0.3311  
## 
## 
## $le_data
## 
## Call:
## lm(formula = Cxcr7 ~ Ahrr, data = this_data)
## 
## Coefficients:
## (Intercept)         Ahrr  
##      1.8324       0.9286</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a>purrr<span class="op">::</span><span class="kw">map</span>(<span class="dt">.x =</span> list_data, </span>
<span id="cb32-2"><a href="#cb32-2"></a>           <span class="dt">.f =</span> lm_Cxcr7_Ahrr) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb32-3"><a href="#cb32-3"></a><span class="st">  </span>purrr<span class="op">::</span><span class="kw">map</span>(<span class="dt">.f =</span> broom<span class="op">::</span>tidy)</span></code></pre></div>
<pre><code>## $clean_data
## # A tibble: 2 x 5
##   term        estimate std.error statistic  p.value
##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1 (Intercept)    2.24     0.195       11.5 1.78e-22
## 2 Ahrr           0.605    0.0589      10.3 3.75e-19
## 
## $hw_data
## # A tibble: 2 x 5
##   term        estimate std.error statistic  p.value
##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1 (Intercept)    2.58     0.134      19.2  3.02e-32
## 2 Ahrr           0.331    0.0412      8.03 5.56e-12
## 
## $le_data
## # A tibble: 2 x 5
##   term        estimate std.error statistic  p.value
##   &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1 (Intercept)    1.83     0.260       7.05 1.22e- 9
## 2 Ahrr           0.929    0.0771     12.0  2.07e-18</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a>gene_long =<span class="st"> </span>clean_data <span class="op">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2"></a><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">pivot_longer</span>(</span>
<span id="cb34-3"><a href="#cb34-3"></a>    <span class="dt">cols =</span> <span class="op">-</span><span class="kw">c</span>(<span class="st">&quot;strain&quot;</span>, <span class="st">&quot;Cxcr7&quot;</span>, <span class="st">&quot;sample&quot;</span>),</span>
<span id="cb34-4"><a href="#cb34-4"></a>    <span class="dt">names_to =</span> <span class="st">&quot;gene_symbols&quot;</span>,</span>
<span id="cb34-5"><a href="#cb34-5"></a>    <span class="dt">values_to =</span> <span class="st">&quot;gene_expression&quot;</span>)</span>
<span id="cb34-6"><a href="#cb34-6"></a></span>
<span id="cb34-7"><a href="#cb34-7"></a>gene_long</span></code></pre></div>
<pre><code>## # A tibble: 8,162 x 5
##    strain sample Cxcr7 gene_symbols gene_expression
##    &lt;chr&gt;  &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;                  &lt;dbl&gt;
##  1 HW     HW1     3.16 Ahrr                    2.57
##  2 HW     HW1     3.16 Aldh3a1                 1.65
##  3 HW     HW1     3.16 Bbs2                    3.52
##  4 HW     HW1     3.16 Ccbl1                   5.47
##  5 HW     HW1     3.16 Col18a1                 7.43
##  6 HW     HW1     3.16 Cyb5a                  10.1 
##  7 HW     HW1     3.16 Cyp1a1                  2.90
##  8 HW     HW1     3.16 Cyp1a2                 10.6 
##  9 HW     HW1     3.16 Cyp1b1                  1.87
## 10 HW     HW1     3.16 Cyp2a1                  8.40
## # … with 8,152 more rows</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a>gene_nest =<span class="st"> </span>gene_long <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb36-2"><a href="#cb36-2"></a><span class="st">  </span><span class="kw">group_by</span>(strain, gene_symbols) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb36-3"><a href="#cb36-3"></a><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">nest</span>()</span>
<span id="cb36-4"><a href="#cb36-4"></a></span>
<span id="cb36-5"><a href="#cb36-5"></a>gene_nest<span class="op">$</span>data[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</span></code></pre></div>
<pre><code>## &lt;list_of&lt;
##   tbl_df&lt;
##     sample         : character
##     Cxcr7          : double
##     gene_expression: double
##   &gt;
## &gt;[2]&gt;
## [[1]]
## # A tibble: 85 x 3
##    sample Cxcr7 gene_expression
##    &lt;chr&gt;  &lt;dbl&gt;           &lt;dbl&gt;
##  1 HW1     3.16           2.57 
##  2 HW2     3.30           0.352
##  3 HW3     2.74           0.254
##  4 HW4     4.22           2.30 
##  5 HW5     3.53           2.02 
##  6 HW6     3.70           1.70 
##  7 WW45    3.62           2.90 
##  8 WW46    3.73           3.11 
##  9 WW47    4.31           3.36 
## 10 WW48    3.85           2.04 
## # … with 75 more rows
## 
## [[2]]
## # A tibble: 85 x 3
##    sample Cxcr7 gene_expression
##    &lt;chr&gt;  &lt;dbl&gt;           &lt;dbl&gt;
##  1 HW1     3.16            1.65
##  2 HW2     3.30            2.14
##  3 HW3     2.74            2.05
##  4 HW4     4.22            5.73
##  5 HW5     3.53            4.18
##  6 HW6     3.70            4.40
##  7 WW45    3.62            7.98
##  8 WW46    3.73            7.01
##  9 WW47    4.31            7.73
## 10 WW48    3.85            7.96
## # … with 75 more rows</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a>lm_Cxcr7_ge =<span class="st"> </span><span class="cf">function</span>(this_data){</span>
<span id="cb38-2"><a href="#cb38-2"></a>  <span class="kw">lm</span>(Cxcr7 <span class="op">~</span><span class="st"> </span>gene_expression, <span class="dt">data =</span> this_data)</span>
<span id="cb38-3"><a href="#cb38-3"></a>}</span>
<span id="cb38-4"><a href="#cb38-4"></a></span>
<span id="cb38-5"><a href="#cb38-5"></a>gene_model =<span class="st"> </span>gene_nest <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb38-6"><a href="#cb38-6"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(<span class="dt">lm_obj =</span> purrr<span class="op">::</span><span class="kw">map</span>(<span class="dt">.x =</span> data, </span>
<span id="cb38-7"><a href="#cb38-7"></a>                                    <span class="dt">.f =</span> lm_Cxcr7_ge),</span>
<span id="cb38-8"><a href="#cb38-8"></a>                <span class="dt">lm_tidy =</span> purrr<span class="op">::</span><span class="kw">map</span>(<span class="dt">.x =</span> lm_obj,</span>
<span id="cb38-9"><a href="#cb38-9"></a>                                     <span class="dt">.f =</span> broom<span class="op">::</span>tidy))</span>
<span id="cb38-10"><a href="#cb38-10"></a></span>
<span id="cb38-11"><a href="#cb38-11"></a></span>
<span id="cb38-12"><a href="#cb38-12"></a></span>
<span id="cb38-13"><a href="#cb38-13"></a>gene_model_long =<span class="st"> </span>gene_model <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb38-14"><a href="#cb38-14"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span>data, <span class="op">-</span>lm_obj) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb38-15"><a href="#cb38-15"></a><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">unnest</span>(lm_tidy) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb38-16"><a href="#cb38-16"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(strain, gene_symbols, term, estimate) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb38-17"><a href="#cb38-17"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(term <span class="op">==</span><span class="st"> &quot;gene_expression&quot;</span>)</span>
<span id="cb38-18"><a href="#cb38-18"></a></span>
<span id="cb38-19"><a href="#cb38-19"></a></span>
<span id="cb38-20"><a href="#cb38-20"></a>gene_model_long</span></code></pre></div>
<pre><code>## # A tibble: 106 x 4
## # Groups:   strain, gene_symbols [106]
##    strain gene_symbols term            estimate
##    &lt;chr&gt;  &lt;chr&gt;        &lt;chr&gt;              &lt;dbl&gt;
##  1 HW     Ahrr         gene_expression    0.331
##  2 HW     Aldh3a1      gene_expression    0.121
##  3 HW     Bbs2         gene_expression    0.616
##  4 HW     Ccbl1        gene_expression    0.573
##  5 HW     Col18a1      gene_expression    0.300
##  6 HW     Cyb5a        gene_expression    0.760
##  7 HW     Cyp1a1       gene_expression    0.115
##  8 HW     Cyp1a2       gene_expression    0.307
##  9 HW     Cyp1b1       gene_expression    0.140
## 10 HW     Cyp2a1       gene_expression    0.742
## # … with 96 more rows</code></pre>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a>gene_model_wide =<span class="st"> </span>gene_model_long <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb40-2"><a href="#cb40-2"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">ungroup</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb40-3"><a href="#cb40-3"></a><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">pivot_wider</span>(<span class="dt">names_from =</span> <span class="st">&quot;strain&quot;</span>,</span>
<span id="cb40-4"><a href="#cb40-4"></a>                     <span class="dt">values_from =</span> <span class="st">&quot;estimate&quot;</span>)</span>
<span id="cb40-5"><a href="#cb40-5"></a></span>
<span id="cb40-6"><a href="#cb40-6"></a></span>
<span id="cb40-7"><a href="#cb40-7"></a>gene_model_wide <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb40-8"><a href="#cb40-8"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> HW, <span class="dt">y =</span> LE)) <span class="op">+</span></span>
<span id="cb40-9"><a href="#cb40-9"></a><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></span>
<span id="cb40-10"><a href="#cb40-10"></a><span class="st">  </span><span class="kw">geom_abline</span>(<span class="dt">slope =</span> <span class="dv">1</span>, <span class="dt">intercept =</span> <span class="dv">0</span>, <span class="dt">colour =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="session3_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a>gene_model_wide <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb41-2"><a href="#cb41-2"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> HW, <span class="dt">y =</span> LE,</span>
<span id="cb41-3"><a href="#cb41-3"></a>             <span class="dt">label =</span> gene_symbols)) <span class="op">+</span></span>
<span id="cb41-4"><a href="#cb41-4"></a><span class="st">  </span><span class="kw">geom_text</span>() <span class="op">+</span></span>
<span id="cb41-5"><a href="#cb41-5"></a><span class="st">  </span><span class="kw">geom_abline</span>(<span class="dt">slope =</span> <span class="dv">1</span>, <span class="dt">intercept =</span> <span class="dv">0</span>, <span class="dt">colour =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="session3_files/figure-html/unnamed-chunk-11-2.png" width="672" /></p>
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1"></a><span class="kw">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 3.6.1 (2019-07-05)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Mojave 10.14.6
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] visdat_0.5.3    readxl_1.3.1    janitor_1.2.0   forcats_0.4.0  
##  [5] stringr_1.4.0   dplyr_0.8.3     purrr_0.3.3     readr_1.3.1    
##  [9] tidyr_1.0.0     tibble_2.1.3    ggplot2_3.2.1   tidyverse_1.2.1
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.3        lubridate_1.7.4   lattice_0.20-38  
##  [4] assertthat_0.2.1  zeallot_0.1.0     digest_0.6.22    
##  [7] utf8_1.1.4        mime_0.7          R6_2.4.1         
## [10] cellranger_1.1.0  backports_1.1.5   evaluate_0.14    
## [13] httr_1.4.1        pillar_1.4.2      rlang_0.4.1      
## [16] lazyeval_0.2.2    rstudioapi_0.10   DT_0.9           
## [19] rmarkdown_1.16    labeling_0.3      htmlwidgets_1.5.1
## [22] munsell_0.5.0     shiny_1.4.0       broom_0.5.2      
## [25] compiler_3.6.1    httpuv_1.5.2      modelr_0.1.5     
## [28] xfun_0.10         pkgconfig_2.0.3   htmltools_0.4.0  
## [31] tidyselect_0.2.5  fansi_0.4.0       crayon_1.3.4     
## [34] withr_2.1.2       later_1.0.0       grid_3.6.1       
## [37] nlme_3.1-141      jsonlite_1.6      xtable_1.8-4     
## [40] gtable_0.3.0      lifecycle_0.1.0   magrittr_1.5     
## [43] scales_1.1.0      cli_1.1.0         stringi_1.4.3    
## [46] farver_2.0.1      promises_1.1.0    snakecase_0.11.0 
## [49] xml2_1.2.2        ellipsis_0.3.0    generics_0.0.2   
## [52] vctrs_0.2.0       tools_3.6.1       glue_1.3.1       
## [55] hms_0.5.2         crosstalk_1.0.0   fastmap_1.0.1    
## [58] yaml_2.2.0        colorspace_1.4-1  rvest_0.3.4      
## [61] knitr_1.25        haven_2.2.0</code></pre>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
