<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Kevin Wang and Garth Tarr" />

<meta name="date" content="2019-05-11" />

<title>session3</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script defer src="https://pro.fontawesome.com/releases/v5.0.13/js/all.js" integrity="sha384-d84LGg2pm9KhR4mCAs3N29GQ4OYNy+K+FBHX8WhimHpPm86c839++MDABegrZ3gn" crossorigin="anonymous"></script>

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<script src="clipboard.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    background-color: #ffffff;
    color: #a0a0a0;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #a0a0a0;  padding-left: 4px; }
div.sourceCode
  { color: #1f1c1b; background-color: #ffffff; }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span. { color: #1f1c1b; } /* Normal */
code span.al { color: #bf0303; background-color: #f7e6e6; font-weight: bold; } /* Alert */
code span.an { color: #ca60ca; } /* Annotation */
code span.at { color: #0057ae; } /* Attribute */
code span.bn { color: #b08000; } /* BaseN */
code span.bu { color: #644a9b; font-weight: bold; } /* BuiltIn */
code span.cf { color: #1f1c1b; font-weight: bold; } /* ControlFlow */
code span.ch { color: #924c9d; } /* Char */
code span.cn { color: #aa5500; } /* Constant */
code span.co { color: #898887; } /* Comment */
code span.cv { color: #0095ff; } /* CommentVar */
code span.do { color: #607880; } /* Documentation */
code span.dt { color: #0057ae; } /* DataType */
code span.dv { color: #b08000; } /* DecVal */
code span.er { color: #bf0303; text-decoration: underline; } /* Error */
code span.ex { color: #0095ff; font-weight: bold; } /* Extension */
code span.fl { color: #b08000; } /* Float */
code span.fu { color: #644a9b; } /* Function */
code span.im { color: #ff5500; } /* Import */
code span.in { color: #b08000; } /* Information */
code span.kw { color: #1f1c1b; font-weight: bold; } /* Keyword */
code span.op { color: #1f1c1b; } /* Operator */
code span.ot { color: #006e28; } /* Other */
code span.pp { color: #006e28; } /* Preprocessor */
code span.re { color: #0057ae; background-color: #e0e9f8; } /* RegionMarker */
code span.sc { color: #3daee9; } /* SpecialChar */
code span.ss { color: #ff5500; } /* SpecialString */
code span.st { color: #bf0303; } /* String */
code span.va { color: #0057ae; } /* Variable */
code span.vs { color: #bf0303; } /* VerbatimString */
code span.wa { color: #bf0303; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">BioinfoSummer - Tidyverse</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="session1.html">Mon: session 1</a>
</li>
<li>
  <a href="session2.html">Tue: session 2</a>
</li>
<li>
  <a href="session3.html">Tue: session 3</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">session3</h1>
<h4 class="author">Kevin Wang and Garth Tarr</h4>
<h4 class="date">05/11/2019</h4>

</div>


<div id="learning-outcomes-of-this-session" class="section level1">
<h1>Learning outcomes of this session</h1>
<ul>
<li>tibbles are better than data.frames</li>
<li>More plotting with <code>ggplot2</code></li>
<li><code>for</code> loops and <code>purrr::map</code> functions</li>
<li>Modelling with <code>broom</code></li>
</ul>
</div>
<div id="loading-packages" class="section level1">
<h1>Loading packages</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(tidyverse)</span></code></pre></div>
</div>
<div id="loading-cleaned-sample-data" class="section level1">
<h1>Loading cleaned sample data</h1>
<p>This session3 data is almost identical to the <code>wide_sample_ge_data.csv</code> file that we just saved. Except that we actually used all the genes in the database.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>session3_data =<span class="st"> </span><span class="kw">read_csv</span>(<span class="dt">file =</span> <span class="st">&quot;data/session3_data.csv&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(<span class="op">-</span>time, <span class="op">-</span>dose)</span></code></pre></div>
<pre><code>## Parsed with column specification:
## cols(
##   .default = col_double(),
##   strain = col_character(),
##   sample = col_character()
## )</code></pre>
<pre><code>## See spec(...) for full column specifications.</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>clean_data =<span class="st"> </span>session3_data <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">mutate_if</span>(is.numeric, <span class="dt">.funs =</span> log)</span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a>clean_data</span></code></pre></div>
<pre><code>## # A tibble: 154 x 56
##    strain sample  Ahrr Aldh3a1  Bbs2 Ccbl1 Cxcr7 Col18a1 Cyb5a Cyp1a1
##    &lt;chr&gt;  &lt;chr&gt;  &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;
##  1 HW     HW1    2.57     1.65  3.52  5.47  3.16    7.43 10.1    2.90
##  2 HW     HW2    0.352    2.14  3.82  5.49  3.30    7.56 10.2    4.22
##  3 HW     HW3    0.254    2.05  3.47  5.17  2.74    6.82 10.1    3.47
##  4 HW     HW4    2.30     5.73  4.24  5.62  4.22    7.77 10.4    9.20
##  5 HW     HW5    2.02     4.18  3.97  5.76  3.53    7.43 10.2    8.84
##  6 HW     HW6    1.70     4.40  3.87  5.50  3.70    7.78 10.00   8.88
##  7 HW     WW45   2.90     7.98  4.13  6.34  3.62    7.19  9.90  10.7 
##  8 HW     WW46   3.11     7.01  4.30  6.26  3.73    7.05  9.85  10.3 
##  9 HW     WW47   3.36     7.73  3.97  6.48  4.31    7.49  9.83  10.7 
## 10 HW     WW48   2.04     7.96  3.74  6.26  3.85    6.90  9.67  10.3 
## # … with 144 more rows, and 46 more variables: Cyp1a2 &lt;dbl&gt;, Cyp1b1 &lt;dbl&gt;,
## #   Cyp2a1 &lt;dbl&gt;, Cyp2s1 &lt;dbl&gt;, Eef1a1 &lt;dbl&gt;, Eml4 &lt;dbl&gt;, Exoc3 &lt;dbl&gt;,
## #   Fdft1 &lt;dbl&gt;, Fmo1 &lt;dbl&gt;, Gapdh &lt;dbl&gt;, Gfer &lt;dbl&gt;, Ghr &lt;dbl&gt;,
## #   Glud1 &lt;dbl&gt;, Hagh &lt;dbl&gt;, Hprt1 &lt;dbl&gt;, Hsd17b12 &lt;dbl&gt;, Ivns1abp &lt;dbl&gt;,
## #   Lasp1 &lt;dbl&gt;, Uvrag &lt;dbl&gt;, Inmt &lt;dbl&gt;, Pbld &lt;dbl&gt;, Mug1 &lt;dbl&gt;,
## #   Neu1 &lt;dbl&gt;, Nfe2l2 &lt;dbl&gt;, Nqo1 &lt;dbl&gt;, Nrbp2 &lt;dbl&gt;, Nudt4 &lt;dbl&gt;,
## #   Pde2a &lt;dbl&gt;, Perp &lt;dbl&gt;, Pgk1 &lt;dbl&gt;, Pmm1 &lt;dbl&gt;, Ppia &lt;dbl&gt;,
## #   Psmb4 &lt;dbl&gt;, Pomp &lt;dbl&gt;, Derl1 &lt;dbl&gt;, Ern1 &lt;dbl&gt;, Rpl13a &lt;dbl&gt;,
## #   Scn4b &lt;dbl&gt;, Sdc1 &lt;dbl&gt;, Sdha &lt;dbl&gt;, Slco1a1 &lt;dbl&gt;, Srxn1 &lt;dbl&gt;,
## #   Tiparp &lt;dbl&gt;, Tpm1 &lt;dbl&gt;, Ugt1a1 &lt;dbl&gt;, Zfp422 &lt;dbl&gt;</code></pre>
</div>
<div id="t-test" class="section level1">
<h1>t-test</h1>
<p>A typical task in bioinformatics is to perform differential gene expression analysis. That is, we want to identify genes that has a different expression patterns between two conditions. While there are many ways that this can be done, we will opt for the simplest approach of using a t-test.</p>
<p>In <code>R</code>, t-test can be performed using the <code>t.test</code> function. Below, we will perform a t-test on the Ahrrr gene between the two strains of mice.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">t.test</span>(Ahrr <span class="op">~</span><span class="st"> </span>strain, <span class="dt">data =</span> clean_data)</span></code></pre></div>
<pre><code>## 
##  Welch Two Sample t-test
## 
## data:  Ahrr by strain
## t = -0.75737, df = 148.97, p-value = 0.45
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -0.5485616  0.2445692
## sample estimates:
## mean in group HW mean in group LE 
##         2.998168         3.150164</code></pre>
<p>However, the output of the <code>t.test</code> function can be a bit hard to read or manipulate into a data.frame. This is where the <code>tidy()</code> function from the <code>broom</code> package can help.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="kw">t.test</span>(Ahrr <span class="op">~</span><span class="st"> </span>strain, <span class="dt">data =</span> clean_data) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="st">  </span>broom<span class="op">::</span><span class="kw">tidy</span>()</span></code></pre></div>
<pre><code>## # A tibble: 1 x 10
##   estimate estimate1 estimate2 statistic p.value parameter conf.low
##      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1   -0.152      3.00      3.15    -0.757   0.450      149.   -0.549
## # … with 3 more variables: conf.high &lt;dbl&gt;, method &lt;chr&gt;,
## #   alternative &lt;chr&gt;</code></pre>
<p>Similarly, we can do the same for the Cxcr7 gene.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw">t.test</span>(Cxcr7 <span class="op">~</span><span class="st"> </span>strain, <span class="dt">data =</span> clean_data) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb11-2"><a href="#cb11-2"></a><span class="st">  </span>broom<span class="op">::</span><span class="kw">tidy</span>()</span></code></pre></div>
<pre><code>## # A tibble: 1 x 10
##   estimate estimate1 estimate2 statistic p.value parameter conf.low
##      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1    -1.19      3.57      4.76     -6.74 1.34e-9      92.9    -1.54
## # … with 3 more variables: conf.high &lt;dbl&gt;, method &lt;chr&gt;,
## #   alternative &lt;chr&gt;</code></pre>
</div>
<div id="kevin-introduce-for-loop-here" class="section level1">
<h1>(Kevin:) introduce for loop here?</h1>
</div>
<div id="doing-multiple-t-tests" class="section level1">
<h1>Doing multiple t-tests</h1>
<p>In the <code>clean_data</code>, we have 54 genes in total. If we want to perform t-test on each gene, should we type out 54 lines of code?</p>
<p>The answer is obviously no. Because such practice is very time inefficient and not generalisable if we receive another dataset in the future.</p>
<p>Now, let’s think about this very difficult task (in a more abstract level). First, we simply run the <code>t.test</code> and <code>broom::tidy</code> function on gene expression values for a selected gene. Then, we do exactly the same operation for another 53 genes. So the whole task is repetitive and we really are apply the following function to 54 datasets (each dataset only contains one selected gene).</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>tidy_test =<span class="st"> </span><span class="cf">function</span>(this_data){</span>
<span id="cb13-2"><a href="#cb13-2"></a>  <span class="kw">t.test</span>(gene_expression <span class="op">~</span><span class="st"> </span>strain, <span class="dt">data =</span> this_data) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb13-3"><a href="#cb13-3"></a><span class="st">    </span>broom<span class="op">::</span><span class="kw">tidy</span>()</span>
<span id="cb13-4"><a href="#cb13-4"></a>}</span></code></pre></div>
<p>So now, our very difficult task of performing 54 t-tests is simplified to spliting <code>clean_data</code> into 54 datasets, one for each gene.</p>
<p>While there are many ways this can be done in <code>R</code>, we will use a <code>tidyverse</code> framework that has the advantage of readable codes and generalisability.</p>
<p>First, let’s pivot our data into a long format.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>gene_long =<span class="st"> </span>clean_data <span class="op">%&gt;%</span></span>
<span id="cb14-2"><a href="#cb14-2"></a><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">pivot_longer</span>(</span>
<span id="cb14-3"><a href="#cb14-3"></a>    <span class="dt">cols =</span> <span class="op">-</span><span class="kw">c</span>(<span class="st">&quot;strain&quot;</span>, <span class="st">&quot;sample&quot;</span>),</span>
<span id="cb14-4"><a href="#cb14-4"></a>    <span class="dt">names_to =</span> <span class="st">&quot;gene_symbol&quot;</span>,</span>
<span id="cb14-5"><a href="#cb14-5"></a>    <span class="dt">values_to =</span> <span class="st">&quot;gene_expression&quot;</span>)</span>
<span id="cb14-6"><a href="#cb14-6"></a></span>
<span id="cb14-7"><a href="#cb14-7"></a>gene_long</span></code></pre></div>
<pre><code>## # A tibble: 8,316 x 4
##    strain sample gene_symbol gene_expression
##    &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;                 &lt;dbl&gt;
##  1 HW     HW1    Ahrr                   2.57
##  2 HW     HW1    Aldh3a1                1.65
##  3 HW     HW1    Bbs2                   3.52
##  4 HW     HW1    Ccbl1                  5.47
##  5 HW     HW1    Cxcr7                  3.16
##  6 HW     HW1    Col18a1                7.43
##  7 HW     HW1    Cyb5a                 10.1 
##  8 HW     HW1    Cyp1a1                 2.90
##  9 HW     HW1    Cyp1a2                10.6 
## 10 HW     HW1    Cyp1b1                 1.87
## # … with 8,306 more rows</code></pre>
<p>Then we will use the <code>group_by</code> function and then apply the <code>nest</code> function from the <code>tidyr</code> package.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a>gene_nest =<span class="st"> </span>gene_long <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb16-2"><a href="#cb16-2"></a><span class="st">  </span><span class="kw">group_by</span>(gene_symbol) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb16-3"><a href="#cb16-3"></a><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">nest</span>(<span class="dt">.key =</span> <span class="st">&quot;small_data&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb16-4"><a href="#cb16-4"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">ungroup</span>()</span></code></pre></div>
<pre><code>## Warning: `.key` is deprecated</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a>gene_nest</span></code></pre></div>
<pre><code>## # A tibble: 54 x 2
##    gene_symbol     small_data
##    &lt;chr&gt;       &lt;list&lt;df[,3]&gt;&gt;
##  1 Ahrr             [154 × 3]
##  2 Aldh3a1          [154 × 3]
##  3 Bbs2             [154 × 3]
##  4 Ccbl1            [154 × 3]
##  5 Cxcr7            [154 × 3]
##  6 Col18a1          [154 × 3]
##  7 Cyb5a            [154 × 3]
##  8 Cyp1a1           [154 × 3]
##  9 Cyp1a2           [154 × 3]
## 10 Cyp1b1           [154 × 3]
## # … with 44 more rows</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a>gene_nest<span class="op">$</span>small_data[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</span></code></pre></div>
<pre><code>## &lt;list_of&lt;
##   tbl_df&lt;
##     strain         : character
##     sample         : character
##     gene_expression: double
##   &gt;
## &gt;[2]&gt;
## [[1]]
## # A tibble: 154 x 3
##    strain sample gene_expression
##    &lt;chr&gt;  &lt;chr&gt;            &lt;dbl&gt;
##  1 HW     HW1              2.57 
##  2 HW     HW2              0.352
##  3 HW     HW3              0.254
##  4 HW     HW4              2.30 
##  5 HW     HW5              2.02 
##  6 HW     HW6              1.70 
##  7 HW     WW45             2.90 
##  8 HW     WW46             3.11 
##  9 HW     WW47             3.36 
## 10 HW     WW48             2.04 
## # … with 144 more rows
## 
## [[2]]
## # A tibble: 154 x 3
##    strain sample gene_expression
##    &lt;chr&gt;  &lt;chr&gt;            &lt;dbl&gt;
##  1 HW     HW1               1.65
##  2 HW     HW2               2.14
##  3 HW     HW3               2.05
##  4 HW     HW4               5.73
##  5 HW     HW5               4.18
##  6 HW     HW6               4.40
##  7 HW     WW45              7.98
##  8 HW     WW46              7.01
##  9 HW     WW47              7.73
## 10 HW     WW48              7.96
## # … with 144 more rows</code></pre>
<p>So, a lot of things happened here. Let’s explain.</p>
<p>You might remember the <code>group_by</code> function when we were trying to use the <code>summarise</code> function to calculate summary statistics. This is exactly what we want to achieve with <code>gene_nest</code>.</p>
<p>Each gene had 154 observations (number rows of <code>clean_data</code>), but in <code>gene_nest</code>, each gene only has “one observation”, which corresponds to a row in the <code>small_data</code> column. <strong>Each row of the <code>small_data</code> column is a <code>data.frame</code> of size 154 rows and 3 columns</strong> (indicated by the <code>[154 x 3]</code>). In each of the <code>[154 x 3]</code> data.frame, you will see three columns (all the columns that we didn’t put into <code>group_by</code>).</p>
<p>Remember, our original aim is to split the <code>gene_long</code> data into 54 smaller data, and we have achived that here with the <code>small_data</code> column.</p>
<p>Now, we simply need to apply the <code>tidy_test</code> function onto the <code>small_data</code> column in this <code>gene_nest</code> data.frame.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">54</span>){</span>
<span id="cb22-2"><a href="#cb22-2"></a>  <span class="kw">tidy_test</span>(gene_nest<span class="op">$</span>small_data[[i]])</span>
<span id="cb22-3"><a href="#cb22-3"></a>}</span></code></pre></div>
</div>
<div id="writing-iterations" class="section level1">
<h1>Writing iterations</h1>
<div id="for-loop" class="section level2">
<h2><code>for</code> loop</h2>
<p><code>for</code> loops are common in any programming language. It simply list out the instructions and the iterations are specified using an index. For example, the following code prints out 1 to 5.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>){</span>
<span id="cb23-2"><a href="#cb23-2"></a>  <span class="kw">print</span>(i)</span>
<span id="cb23-3"><a href="#cb23-3"></a>}</span></code></pre></div>
<pre><code>## [1] 1
## [1] 2
## [1] 3
## [1] 4
## [1] 5</code></pre>
<p>We can use the <code>for</code> loop to compute the <code>tidy_test</code> output for all 54 genes.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">54</span>){</span>
<span id="cb25-2"><a href="#cb25-2"></a>  <span class="kw">tidy_test</span>(gene_nest<span class="op">$</span>small_data[[i]])</span>
<span id="cb25-3"><a href="#cb25-3"></a>}</span></code></pre></div>
<p>However, the output of this loop is a <code>list</code> object requiring further manipulations to make it suitable for visualisation purposes. Hence, we will learn a new way for plotting.</p>
</div>
<div id="map-function-in-purrr" class="section level2">
<h2><code>map</code> function in <code>purrr</code></h2>
<p>The <code>map</code> function from the <code>purrr</code> package is a tremendously powerful function. In simply terms, it allows us to write a <code>for</code> loop is a very simply and readable way.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>purrr<span class="op">::</span><span class="kw">map</span>(<span class="dt">.x =</span> gene_nest<span class="op">$</span>small_data,</span>
<span id="cb26-2"><a href="#cb26-2"></a>           <span class="dt">.f =</span> tidy_test)</span></code></pre></div>
<p>Why is this code better than the <code>for</code> loop even though they achieved the same output? The biggest reason is that <strong>we don’t have to read and understand individual lines</strong> inside the <code>for</code> loop. We simply need to know:</p>
<ul>
<li><code>.x</code>, the input</li>
<li><code>.f</code>, the function</li>
</ul>
<p>Then the function <code>.f</code> will be applied onto every element of the <code>.x</code> input.</p>
<p>Another reason why <code>map</code> function is superior to a <code>for</code> loop is how easy it can work with the <code>mutate</code> function in <code>dplyr</code>. The following code can be read as:</p>
<ul>
<li>In the <code>gene_nest</code> data, we want to <code>mutate</code> (create) a new column</li>
<li>This new column will be called <code>tidy_result</code> and it is defined as mapping the <code>tidy_test</code> function to every element if the <code>small_data</code> column.</li>
</ul>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>gene_test =<span class="st"> </span>gene_nest <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb27-2"><a href="#cb27-2"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(</span>
<span id="cb27-3"><a href="#cb27-3"></a>    <span class="dt">tidy_result =</span> purrr<span class="op">::</span><span class="kw">map</span>(</span>
<span id="cb27-4"><a href="#cb27-4"></a>      <span class="dt">.x =</span> small_data,</span>
<span id="cb27-5"><a href="#cb27-5"></a>      <span class="dt">.f =</span> tidy_test)</span>
<span id="cb27-6"><a href="#cb27-6"></a>  )</span>
<span id="cb27-7"><a href="#cb27-7"></a></span>
<span id="cb27-8"><a href="#cb27-8"></a>gene_test</span></code></pre></div>
<pre><code>## # A tibble: 54 x 3
##    gene_symbol     small_data tidy_result      
##    &lt;chr&gt;       &lt;list&lt;df[,3]&gt;&gt; &lt;list&gt;           
##  1 Ahrr             [154 × 3] &lt;tibble [1 × 10]&gt;
##  2 Aldh3a1          [154 × 3] &lt;tibble [1 × 10]&gt;
##  3 Bbs2             [154 × 3] &lt;tibble [1 × 10]&gt;
##  4 Ccbl1            [154 × 3] &lt;tibble [1 × 10]&gt;
##  5 Cxcr7            [154 × 3] &lt;tibble [1 × 10]&gt;
##  6 Col18a1          [154 × 3] &lt;tibble [1 × 10]&gt;
##  7 Cyb5a            [154 × 3] &lt;tibble [1 × 10]&gt;
##  8 Cyp1a1           [154 × 3] &lt;tibble [1 × 10]&gt;
##  9 Cyp1a2           [154 × 3] &lt;tibble [1 × 10]&gt;
## 10 Cyp1b1           [154 × 3] &lt;tibble [1 × 10]&gt;
## # … with 44 more rows</code></pre>
<p>This is a very powerful framework, because it allows us to create new statistics horizonally inside a <code>data.frame</code>, keep in track of the matching genes/observations. You may not realise this, but one of the most common mistake in data analysis is mis-matching of rows when combining datasets and statistics. The use of the <code>mutate</code> and <code>map</code> functions allows all intermediate steps to be stored as a column inside a data.frame, precenting any mis-matchings.</p>
</div>
</div>
<div id="plotting-t-test-results" class="section level1">
<h1>Plotting t-test results</h1>
<p>The <code>tidy_result</code> column in <code>gene_test</code> is now a list of tibbles (think of them as data.frames). Each of these tibbles contains t-test statistic and p-value associated with each gene. We will now need to expand this column. The function we will use is <code>unnest</code> from the <code>tidyr</code> package. As you may already guess, this function undo the <code>nest</code> operation. In this case, we want to undo the <code>tidy_result</code> column only.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1"></a>gene_test<span class="op">$</span>tidy_result[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]</span></code></pre></div>
<pre><code>## [[1]]
## # A tibble: 1 x 10
##   estimate estimate1 estimate2 statistic p.value parameter conf.low
##      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1   -0.152      3.00      3.15    -0.757   0.450      149.   -0.549
## # … with 3 more variables: conf.high &lt;dbl&gt;, method &lt;chr&gt;,
## #   alternative &lt;chr&gt;
## 
## [[2]]
## # A tibble: 1 x 10
##   estimate estimate1 estimate2 statistic p.value parameter conf.low
##      &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1   -0.494      6.36      6.86    -0.959   0.339      145.    -1.51
## # … with 3 more variables: conf.high &lt;dbl&gt;, method &lt;chr&gt;,
## #   alternative &lt;chr&gt;</code></pre>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a>gene_test_long =<span class="st"> </span>gene_test <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb31-2"><a href="#cb31-2"></a><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">unnest</span>(tidy_result)</span>
<span id="cb31-3"><a href="#cb31-3"></a></span>
<span id="cb31-4"><a href="#cb31-4"></a>gene_test_long</span></code></pre></div>
<pre><code>## # A tibble: 54 x 12
##    gene_symbol small_data estimate estimate1 estimate2 statistic p.value
##    &lt;chr&gt;       &lt;list&lt;df[&gt;    &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt;
##  1 Ahrr         [154 × 3]  -0.152       3.00      3.15    -0.757 4.50e-1
##  2 Aldh3a1      [154 × 3]  -0.494       6.36      6.86    -0.959 3.39e-1
##  3 Bbs2         [154 × 3]  -0.264       4.18      4.44    -3.25  1.44e-3
##  4 Ccbl1        [154 × 3]  -0.0702      6.25      6.32    -0.640 5.23e-1
##  5 Cxcr7        [154 × 3]  -1.19        3.57      4.76    -6.74  1.34e-9
##  6 Col18a1      [154 × 3]  -0.0752      7.27      7.35    -1.03  3.02e-1
##  7 Cyb5a        [154 × 3]   0.0783     10.2      10.1      1.48  1.40e-1
##  8 Cyp1a1       [154 × 3]  -0.485       8.89      9.38    -1.03  3.07e-1
##  9 Cyp1a2       [154 × 3]  -0.216      12.0      12.2     -1.17  2.42e-1
## 10 Cyp1b1       [154 × 3]  -0.502       4.84      5.35    -1.11  2.68e-1
## # … with 44 more rows, and 5 more variables: parameter &lt;dbl&gt;,
## #   conf.low &lt;dbl&gt;, conf.high &lt;dbl&gt;, method &lt;chr&gt;, alternative &lt;chr&gt;</code></pre>
<p>Armed with all the t-test statistics and p-values, we may want to perform some quick visualisations.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a>gene_test_long <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb33-2"><a href="#cb33-2"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> p.value)) <span class="op">+</span></span>
<span id="cb33-3"><a href="#cb33-3"></a><span class="st">  </span><span class="kw">geom_histogram</span>()</span></code></pre></div>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="session3_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<p>This histogram shows that we actually have about 1/3 of genes significantly differentially expressed (DE) between the two strains. This is usually a problem with gene expression analysis because if all t-test assumptions are met, the p-values should be uniformly distributed over the [0,1] interval. This inflation is could be attributed to normalisation issues with the data since we have only taken the most naive approach of taking the logarithm. Additionally, we should also note that this is a <em>targeted</em> gene expression data where genes are typically selected due to differential patterns they showed in pilot studies. We will not commenton the biological implication of this histogram further, but proceed with another common analysis, which is to create boxplots of the top differentially expressed genes.</p>
<div id="boxplots-of-top-differentially-expressed-genes" class="section level2">
<h2>Boxplots of top differentially expressed genes</h2>
<p>In this section, we will first extract the top 6 DE genes and make a boxplot for each of them between the two strains.</p>
<p>The following code allows us to subset the data.frame to the 6 genes with smallest p-values (i.e. most differentially expressed). However, we shouldn’t forget that all the gene expression values are still stored in the <code>small_data</code> column. We need to apply the <code>unnest</code> operation to extract these values.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a>gene_test_long <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb35-2"><a href="#cb35-2"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="kw">rank</span>(p.value) <span class="op">&lt;=</span><span class="st"> </span><span class="dv">6</span>) </span></code></pre></div>
<pre><code>## # A tibble: 6 x 12
##   gene_symbol small_data estimate estimate1 estimate2 statistic  p.value
##   &lt;chr&gt;       &lt;list&lt;df[&gt;    &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
## 1 Cxcr7        [154 × 3]   -1.19       3.57      4.76     -6.74 1.34e- 9
## 2 Cyp2a1       [154 × 3]   -0.560      8.71      9.27     -4.93 2.43e- 6
## 3 Mug1         [154 × 3]   -0.779      9.91     10.7      -8.34 5.71e-14
## 4 Nqo1         [154 × 3]   -1.21       7.00      8.21     -5.93 2.01e- 8
## 5 Derl1        [154 × 3]   -0.242      6.82      7.07     -4.69 6.15e- 6
## 6 Tpm1         [154 × 3]   -0.339      3.62      3.96     -5.06 1.23e- 6
## # … with 5 more variables: parameter &lt;dbl&gt;, conf.low &lt;dbl&gt;,
## #   conf.high &lt;dbl&gt;, method &lt;chr&gt;, alternative &lt;chr&gt;</code></pre>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a>boxplot_data =<span class="st"> </span>gene_test_long <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb37-2"><a href="#cb37-2"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(<span class="kw">rank</span>(p.value) <span class="op">&lt;=</span><span class="st"> </span><span class="dv">6</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb37-3"><a href="#cb37-3"></a><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">unnest</span>(small_data)</span>
<span id="cb37-4"><a href="#cb37-4"></a></span>
<span id="cb37-5"><a href="#cb37-5"></a>boxplot_data</span></code></pre></div>
<pre><code>## # A tibble: 924 x 14
##    gene_symbol strain sample gene_expression estimate estimate1 estimate2
##    &lt;chr&gt;       &lt;chr&gt;  &lt;chr&gt;            &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
##  1 Cxcr7       HW     HW1               3.16    -1.19      3.57      4.76
##  2 Cxcr7       HW     HW2               3.30    -1.19      3.57      4.76
##  3 Cxcr7       HW     HW3               2.74    -1.19      3.57      4.76
##  4 Cxcr7       HW     HW4               4.22    -1.19      3.57      4.76
##  5 Cxcr7       HW     HW5               3.53    -1.19      3.57      4.76
##  6 Cxcr7       HW     HW6               3.70    -1.19      3.57      4.76
##  7 Cxcr7       HW     WW45              3.62    -1.19      3.57      4.76
##  8 Cxcr7       HW     WW46              3.73    -1.19      3.57      4.76
##  9 Cxcr7       HW     WW47              4.31    -1.19      3.57      4.76
## 10 Cxcr7       HW     WW48              3.85    -1.19      3.57      4.76
## # … with 914 more rows, and 7 more variables: statistic &lt;dbl&gt;,
## #   p.value &lt;dbl&gt;, parameter &lt;dbl&gt;, conf.low &lt;dbl&gt;, conf.high &lt;dbl&gt;,
## #   method &lt;chr&gt;, alternative &lt;chr&gt;</code></pre>
<p>And now, let’s perform the final plotting!</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a>boxplot_data <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb39-2"><a href="#cb39-2"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> strain, </span>
<span id="cb39-3"><a href="#cb39-3"></a>             <span class="dt">y =</span> gene_expression)) <span class="op">+</span></span>
<span id="cb39-4"><a href="#cb39-4"></a><span class="st">  </span><span class="kw">geom_boxplot</span>() <span class="op">+</span></span>
<span id="cb39-5"><a href="#cb39-5"></a><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>gene_symbol, <span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>)</span></code></pre></div>
<p><img src="session3_files/figure-html/unnamed-chunk-17-1.png" width="672" /></p>
<p>We can also prettify the plots by creating a <code>label</code> column, which combines the gene symbols and p-values.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a>boxplot_data <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb40-2"><a href="#cb40-2"></a><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(<span class="dt">label =</span> <span class="kw">paste0</span>(</span>
<span id="cb40-3"><a href="#cb40-3"></a>    gene_symbol, <span class="st">&quot;, p-value = &quot;</span>, <span class="kw">signif</span>(p.value, <span class="dv">2</span>))) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb40-4"><a href="#cb40-4"></a><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> strain, </span>
<span id="cb40-5"><a href="#cb40-5"></a>             <span class="dt">y =</span> gene_expression)) <span class="op">+</span></span>
<span id="cb40-6"><a href="#cb40-6"></a><span class="st">  </span><span class="kw">geom_boxplot</span>() <span class="op">+</span></span>
<span id="cb40-7"><a href="#cb40-7"></a><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>label, <span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>) <span class="op">+</span></span>
<span id="cb40-8"><a href="#cb40-8"></a><span class="st">  </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&quot;Strains&quot;</span>, </span>
<span id="cb40-9"><a href="#cb40-9"></a>       <span class="dt">y =</span> <span class="st">&quot;Gene expression values&quot;</span>)</span></code></pre></div>
<p><img src="session3_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<!-- # Regression analysis (extension) -->
<!-- ```{r} -->
<!-- hw_data = clean_data %>%  -->
<!--   dplyr::filter(strain == "HW") -->
<!-- le_data = clean_data %>%  -->
<!--   dplyr::filter(strain == "LE") -->
<!-- clean_data %>%  -->
<!--   ggplot(aes(x = Ahrr, -->
<!--              y = Cxcr7, -->
<!--              colour = strain)) + -->
<!--   geom_point() + -->
<!--   geom_smooth(method = "lm") -->
<!-- lm(Cxcr7 ~ Ahrr, data = clean_data) -->
<!-- lm(Cxcr7 ~ Ahrr, data = hw_data) -->
<!-- lm(Cxcr7 ~ Ahrr, data = le_data) -->
<!-- lm_Cxcr7_Ahrr = function(this_data){ -->
<!--   lm(Cxcr7 ~ Ahrr, data = this_data) -->
<!-- } -->
<!-- list_data = list( -->
<!--   clean_data = clean_data,  -->
<!--   hw_data = hw_data, -->
<!--   le_data = le_data) -->
<!-- purrr::map(.x = list_data,  -->
<!--            .f = lm_Cxcr7_Ahrr) -->
<!-- purrr::map(.x = list_data,  -->
<!--            .f = lm_Cxcr7_Ahrr) %>%  -->
<!--   purrr::map(.f = broom::tidy) -->
<!-- ``` -->
<!-- ```{r} -->
<!-- gene_long = clean_data %>% -->
<!--   tidyr::pivot_longer( -->
<!--     cols = -c("strain", "Cxcr7", "sample"), -->
<!--     names_to = "gene_symbol", -->
<!--     values_to = "gene_expression") -->
<!-- gene_long -->
<!-- gene_nest = gene_long %>%  -->
<!--   group_by(strain, gene_symbol) %>%  -->
<!--   tidyr::nest() -->
<!-- gene_nest$data[1:2] -->
<!-- lm_Cxcr7_ge = function(this_data){ -->
<!--   lm(Cxcr7 ~ gene_expression, data = this_data) -->
<!-- } -->
<!-- gene_model = gene_nest %>%  -->
<!--   dplyr::mutate(lm_obj = purrr::map(.x = data,  -->
<!--                                     .f = lm_Cxcr7_ge), -->
<!--                 lm_tidy = purrr::map(.x = lm_obj, -->
<!--                                      .f = broom::tidy)) -->
<!-- gene_model_long = gene_model %>%  -->
<!--   dplyr::select(-data, -lm_obj) %>%  -->
<!--   tidyr::unnest(lm_tidy) %>%  -->
<!--   dplyr::select(strain, gene_symbol, term, estimate) %>%  -->
<!--   dplyr::filter(term == "gene_expression") -->
<!-- gene_model_long -->
<!-- gene_model_wide = gene_model_long %>%  -->
<!--   dplyr::ungroup() %>%  -->
<!--   tidyr::pivot_wider(names_from = "strain", -->
<!--                      values_from = "estimate") -->
<!-- gene_model_wide %>%  -->
<!--   ggplot(aes(x = HW, y = LE)) + -->
<!--   geom_point() + -->
<!--   geom_abline(slope = 1, intercept = 0, colour = "red") -->
<!-- gene_model_wide %>%  -->
<!--   ggplot(aes(x = HW, y = LE, -->
<!--              label = gene_symbols)) + -->
<!--   geom_text() + -->
<!--   geom_abline(slope = 1, intercept = 0, colour = "red") -->
<!-- ``` -->
</div>
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a><span class="kw">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 3.6.1 (2019-07-05)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Mojave 10.14.6
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] visdat_0.5.3    readxl_1.3.1    janitor_1.2.0   forcats_0.4.0  
##  [5] stringr_1.4.0   dplyr_0.8.3     purrr_0.3.3     readr_1.3.1    
##  [9] tidyr_1.0.0     tibble_2.1.3    ggplot2_3.2.1   tidyverse_1.2.1
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.3        lubridate_1.7.4   lattice_0.20-38  
##  [4] assertthat_0.2.1  zeallot_0.1.0     digest_0.6.22    
##  [7] utf8_1.1.4        mime_0.7          R6_2.4.1         
## [10] cellranger_1.1.0  backports_1.1.5   evaluate_0.14    
## [13] httr_1.4.1        pillar_1.4.2      rlang_0.4.1      
## [16] lazyeval_0.2.2    rstudioapi_0.10   DT_0.9           
## [19] rmarkdown_1.16    labeling_0.3      htmlwidgets_1.5.1
## [22] munsell_0.5.0     shiny_1.4.0       broom_0.5.2      
## [25] compiler_3.6.1    httpuv_1.5.2      modelr_0.1.5     
## [28] xfun_0.10         pkgconfig_2.0.3   htmltools_0.4.0  
## [31] tidyselect_0.2.5  fansi_0.4.0       crayon_1.3.4     
## [34] withr_2.1.2       later_1.0.0       grid_3.6.1       
## [37] nlme_3.1-141      jsonlite_1.6      xtable_1.8-4     
## [40] gtable_0.3.0      lifecycle_0.1.0   magrittr_1.5     
## [43] scales_1.1.0      cli_1.1.0         stringi_1.4.3    
## [46] farver_2.0.1      promises_1.1.0    snakecase_0.11.0 
## [49] xml2_1.2.2        ellipsis_0.3.0    generics_0.0.2   
## [52] vctrs_0.2.0       tools_3.6.1       glue_1.3.1       
## [55] hms_0.5.2         crosstalk_1.0.0   fastmap_1.0.1    
## [58] yaml_2.2.0        colorspace_1.4-1  rvest_0.3.4      
## [61] knitr_1.25        haven_2.2.0</code></pre>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
