---
title: "session3"
author: "Kevin Wang and Garth Tarr"
date: "05/11/2019"
output:
  html_document:
    css: style.css
---



# Learning outcomes of this session

+ More plotting with `ggplot2`
+ `for` loops and `purrr::map` functions
+ Modelling with `broom`



# Loading packages

```{r}
library(tidyverse)
```


# Loading cleaned sample data


This session3 data is almost identical to the `wide_sample_ge_data.csv` file that we just saved. Except that we actually used all the genes in the database. 

```{r}
session3_data = read_csv(file = "data/session3_data.csv") %>% 
  dplyr::select(-time, -dose)

clean_data = session3_data %>% 
  dplyr::mutate_if(is.numeric, .funs = log)

clean_data
```


# t-test

A typical task in bioinformatics is to perform differential gene expression analysis. That is, we want to identify genes that has a different expression patterns between two conditions. While there are many ways that this can be done, we will opt for the simplest approach of using a t-test. 

In `R`, t-test can be performed using the `t.test` function. Below, we will perform a t-test on the Ahrrr gene between the two strains of mice. 

```{r}
t.test(Ahrr ~ strain, data = clean_data)
```

However, the output of the `t.test` function can be a bit hard to read or manipulate into a data.frame. This is where the `tidy()` function from the `broom` package can help.

```{r}
t.test(Ahrr ~ strain, data = clean_data) %>% 
  broom::tidy()
```


Similarly, we can do the same for the Cxcr7 gene. 

```{r}
t.test(Cxcr7 ~ strain, data = clean_data) %>% 
  broom::tidy()
```



# Doing multiple t-tests

In the `clean_data`, we have 54 genes in total. If we want to perform t-test on each gene, should we type out 54 lines of code? 

The answer is obviously no. Because such practice is very time inefficient and not generalisable if we receive another dataset in the future. 


Now, let's think about this very difficult task (in a more abstract level). First, we simply run the `t.test` and `broom::tidy` function on gene expression values for a selected gene. Then, we do exactly the same operation for another 53 genes. So the whole task is repetitive and we really are apply the following function to 54 datasets (each dataset only contains one selected gene). 

```{r}
tidy_test = function(this_data){
  t.test(gene_expression ~ strain, data = this_data) %>% 
    broom::tidy()
}
```



So now, our very difficult task of performing 54 t-tests is simplified to spliting `clean_data` into 54 datasets, one for each gene. 

While there are many ways this can be done in `R`, we will use a `tidyverse` framework that has the advantage of readable codes and generalisability. 

First, let's pivot our data into a long format. 

```{r}
gene_long = clean_data %>%
  tidyr::pivot_longer(
    cols = -c("strain", "sample"),
    names_to = "gene_symbols",
    values_to = "gene_expression")

gene_long
```


Then we will use the `group_by` function and then apply the `nest` function from the `tidyr` package. 

```{r}
gene_nest = gene_long %>% 
  group_by(gene_symbols) %>% 
  tidyr::nest()

gene_nest

gene_nest$data[1:2]
```


So, a lot of things happened here. Let's explain. 

You might remember the `group_by` function when we were trying to use the `summarise` function to calculate summary statistics. This is exactly what we want to achieve here with `gene_nest`. Each gene used to have 154 observations (number rows of `clean_data`), but in `gene_nest`, each gene only has "one observation", which corresponds to a row in the `data` column. **Each row of the `data` column is a `data.frame` of size 154 rows and 3 columns**, indicated by the `[154 x 3]`. In each of the `[154 x 3]` data.frame, you will see three columns (all the columns that we didn't put into `group_by`). 


```{r}
gene_test = gene_nest %>% 
  dplyr::mutate(
    tidy_result = purrr::map(
      .x = data,
      .f = tidy_test))

gene_model_long = gene_test %>% 
  tidyr::unnest(tidy_result)


gene_model_long %>% 
  ggplot(aes(x = p.value)) +
  geom_histogram()


gene_model_long %>% 
  dplyr::filter(p.value < 0.0001) %>% 
  tidyr::unnest(data) %>% 
  dplyr::mutate(
    label = paste0(gene_symbols, 
                   ", p-value = ", 
                   signif(p.value, 2))) %>% 
  ggplot(aes(x = strain, 
             y = gene_expression)) +
  geom_boxplot() +
  facet_wrap(~ label, scales = "free")
```



# Regression analysis (extension)

```{r}


hw_data = clean_data %>% 
  dplyr::filter(strain == "HW")

le_data = clean_data %>% 
  dplyr::filter(strain == "LE")


clean_data %>% 
  ggplot(aes(x = Ahrr,
             y = Cxcr7,
             colour = strain)) +
  geom_point() +
  geom_smooth(method = "lm")


lm(Cxcr7 ~ Ahrr, data = clean_data)
lm(Cxcr7 ~ Ahrr, data = hw_data)
lm(Cxcr7 ~ Ahrr, data = le_data)

lm_Cxcr7_Ahrr = function(this_data){
  lm(Cxcr7 ~ Ahrr, data = this_data)
}

list_data = list(
  clean_data = clean_data, 
  hw_data = hw_data,
  le_data = le_data)

purrr::map(.x = list_data, 
           .f = lm_Cxcr7_Ahrr)

purrr::map(.x = list_data, 
           .f = lm_Cxcr7_Ahrr) %>% 
  purrr::map(.f = broom::tidy)
```




```{r}
gene_long = clean_data %>%
  tidyr::pivot_longer(
    cols = -c("strain", "Cxcr7", "sample"),
    names_to = "gene_symbols",
    values_to = "gene_expression")

gene_long

gene_nest = gene_long %>% 
  group_by(strain, gene_symbols) %>% 
  tidyr::nest()

gene_nest$data[1:2]

lm_Cxcr7_ge = function(this_data){
  lm(Cxcr7 ~ gene_expression, data = this_data)
}

gene_model = gene_nest %>% 
  dplyr::mutate(lm_obj = purrr::map(.x = data, 
                                    .f = lm_Cxcr7_ge),
                lm_tidy = purrr::map(.x = lm_obj,
                                     .f = broom::tidy))



gene_model_long = gene_model %>% 
  dplyr::select(-data, -lm_obj) %>% 
  tidyr::unnest(lm_tidy) %>% 
  dplyr::select(strain, gene_symbols, term, estimate) %>% 
  dplyr::filter(term == "gene_expression")


gene_model_long


gene_model_wide = gene_model_long %>% 
  dplyr::ungroup() %>% 
  tidyr::pivot_wider(names_from = "strain",
                     values_from = "estimate")


gene_model_wide %>% 
  ggplot(aes(x = HW, y = LE)) +
  geom_point() +
  geom_abline(slope = 1, intercept = 0, colour = "red")

gene_model_wide %>% 
  ggplot(aes(x = HW, y = LE,
             label = gene_symbols)) +
  geom_text() +
  geom_abline(slope = 1, intercept = 0, colour = "red")
```





# Session Info
```{r}
sessionInfo()
```

