---
title: "Enter the `tidyverse` with R and RStudio"
subtitle: ""
author: "Kevin Wang and Garth Tarr"
date: "2-3 December 2019, Sydney, Australia"
output:
  xaringan::moon_reader:
    css: ["default", "../assets/sydney-fonts.css", "../assets/sydney.css"]
    self_contained: false # if true, fonts will be stored locally
    seal: true # show a title slide with YAML information
    includes:
      in_header: "../assets/mathjax-equation-numbers.html"
    nature:
      beforeInit: ["../assets/remark-zoom.js", "https://platform.twitter.com/widgets.js"]
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: true
      ratio: '16:9' # alternatives '16:9' or '4:3' or others e.g. 13:9
      navigation:
        scroll: false # disable slide transitions by scrolling
---


# Your instructors

+ Kevin: PhD candidate in statistics and bioinformatics.
+ Garth: Lecturer in statistics and data science. 

---

# Ice breaker


---

# Workshop outline: make a boxplot similar to [this](https://www.nature.com/articles/cmi2014102.pdf)

.pull-left-2[
<center>
<img src="../assets/publication_boxplot.png", width="100%">
</center>
]

--

.pull-right-1[

1. Get gene expression data
1. Get clinical data
1. Understand what the data structure
1. Figure out how to join the two data
1. Reshape the data
1. Visualisation
1. Compute t-test p-values and add them to the plots

]

.footnote[
Neve Polimeno, M., Ierano, C., D'Alterio, C., Simona Losito, N., Napolitano, M., Portella, L., ... Scala, S. (2015). CXCR4 expression affects overall survival of HCC patients whereas CXCR7 expression does not. _Cellular & molecular immunology_, 12(4), 474â€“482. [doi:10.1038/cmi.2014.102](https://www.nature.com/articles/cmi2014102.pdf)
]

---

# What is tidy data?

1. Each row is an observation unit
1. Each column is a variable 
1. Each cell is one value

```{r, include = FALSE}
library(tidyverse)
set.seed(123)
tidy_data = readr::read_csv("../data/clean_sample_data.csv")
```

```{r}
sample_n(tidy_data, size = 10)
```

---

# Why tidy data?

+ Because it will make our analyses and visualisation easier. 
+ An example of dirty data:

<center>
<img src="../assets/UAC.png", width="100%">
</center>

---

# Visualisation Gallery

+ NYC squirrel sighting: https://twitter.com/i/status/1189658282618699776
+ Australian election result: https://srkobakian.github.io/sugarbag/index.html

<center>
<img src="https://github.com/srkobakian/sugarbag/raw/master/man/figures/README-animated-1.gif", width="60%">
</center>

---

# A simple bioinformatics data

+ We will be working on a real gene expression data from [this article](https://rnajournal.cshlp.org/content/19/1/51.full): **Systematic evaluation of medium-throughput mRNA abundance platforms** by Prokopec et. al. 

+ The original data was generated to understand reproducibility of three gene expression technologies by comparing gene expression of mice treated with different diet. 

+ There are two pieces of data: 

1. **The sample data** was downloaded from the [journal's website](https://rnajournal.cshlp.org/content/19/1/51/suppl/DC1).

1. **The gene expression data** was downloaded from the [Gene Expression Omnibus](https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE43251&format=file&file=GSE43251%5FNanoString%5Fnon%2Dnormalized%2Etxt%2Egz), accession ID GSE43251.

We will play with the sample data on Monday, combine the two data for visualisation on Tuesday morning and perform basic modelling on Tuesday afternoon. 

---

# Challenges of this data

1. Understanding the data structure and read the two data into `R`.
1. Basic manipulation of the two data. 
1. Merging the two data. 
1. Visualising the data.
1. Reshaping the data for visualisation and modelling purpose.
1. Perform basic modelling on the data and extract insights.

---

# Session 1: reading data and basic manipulations

.pull-left[

> **Why can't I do everything in Excel?**

+ Processing raw data to meaningful scientific result is often complex and iterative. 

+ It is better and more rigorous to have a R-Markdown file documenting this process (reproducible).

+ In this session, we will learn how to read in the raw data, identify some problems with the data file, remove problemative rows and columns of the data and save a cleaned data file.

]

.pull-right[

<img src="../assets/noexcel.png", width="100%">

]

.footnote[
Ziemann, M., Eren, Y. & El-Osta, A. (2016). Gene name errors are widespread in the scientific literature. _Genome Biology_ 17, 177 [doi:10.1186/s13059-016-1044-7](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1044-7)
]

---

# Pipe operator `%>%`

.large[

```
awesome_data =
  raw_interesting_data %>%
  mutate(somehow) %>%
  filter(the_good_parts) %>%
  summarise(required_statistics)
```

Instead of:

```
awesome_data = 
  summarise(
    filter(
      mutate(raw_interesting_data, somehow),
      the_good_parts),
    required_statistics)
```

]

.footnote[
Read more: https://r4ds.had.co.nz/pipes.html
]

---

# Extension exercises of Session 1

1. Return to the `diamonds` data frame from the start of the session.  Find the number of observations for each cut. 
2. Calculate summary statistics of the price for each cut (e.g. mean, median, min, max, ...)

---

# Session 2: visualisation with `ggplot2` 

+ `ggplot2` is rigorous framework for plotting
+ The `gg` in `ggplot2` stands for "grammar of graphics".
+ in order to produce a plot, a strict set of coding structure must be followed. 
+ In particular, **every characteristic of your plot is produced using one variable**. This avoids the "double y-axes" plot:
<center>
<img src="https://statmodeling.stat.columbia.edu/wp-content/uploads/2015/05/journal.pone_.0112042.g001.png", width="40%">
</center>

+ Other articles about bad visualisations can be found [here](https://venngage.com/blog/misleading-graphs/) and [here](https://www.businessinsider.com.au/the-27-worst-charts-of-all-time-2013-6?r=US&IR=T#i-never-thought-it-was-possible-but-i-actually-understand-soccer-less-after-looking-at-this-chart-3).

---

## A quick example of `ggplot2`



```{r, eval = FALSE, echo = FALSE}
library(dslabs)
data(divorce_margarine, package = "dslabs")

divorce_margarine_scaled = divorce_margarine %>% 
  dplyr::transmute(
    year,
    divorce = scale(divorce_rate_maine),
    margarine = scale(margarine_consumption_per_capita)
  )

readr::write_csv(divorce_margarine_scaled, path = "../data/divorce_margarine_scaled.csv")
```

```{r,include=FALSE}
divorce_margarine_scaled = read_csv("../data/divorce_margarine_scaled.csv")
```

.pull-left[
```{r}
divorce_margarine_scaled
```
]


.pull-right[
+ divorce should be mapped to the x-axis
+ margarine should be mapped to the y-axis
]
---

## First taste of `ggplot2`

```{r, fig.height=4}
divorce_margarine_scaled %>% 
  ggplot() #<<
```

---

## First taste of `ggplot2`

```{r, fig.height=4}
divorce_margarine_scaled %>% 
  ggplot(aes(
    x = divorce, #<<
    y = margarine #<<
  ))
```

---

## First taste of `ggplot2`

```{r, fig.height=4}
divorce_margarine_scaled %>% 
  ggplot(aes(
    x = divorce,
    y = margarine
  )) +
  geom_point() #<<
```

---

## First taste of `ggplot2`

```{r, fig.height=4}
divorce_margarine_scaled %>% 
  ggplot(aes(
    x = divorce,
    y = margarine,
    label = year #<<
  )) +
  geom_point() +
  geom_text() #<<
```

---

## First taste of `ggplot2`

```{r, fig.height=4}
divorce_margarine_scaled %>% 
  ggplot(aes(
    x = divorce,
    y = margarine,
    label = year
  )) +
  geom_point() +
  geom_text() +
  labs(title = "Divorce rate and margarine consumption") #<<
```

---

## Tidy is goal-dependent: treating each year as independent

```{r, fig.height=4}
divorce_margarine_scaled %>% 
  ggplot(aes(
    x = divorce,
    y = margarine,
    label = year
  )) +
  geom_point() +
  geom_text() +
  labs(title = "Divorce rate and margarine consumption")
```

---

## Can you make this plot using the data above?
.pull-left[

```{r}
divorce_margarine_scaled
```
]

.pull-right[
```{r, fig.height=6, echo = FALSE}
long_data = divorce_margarine_scaled %>%
  tidyr::pivot_longer(
    cols = -c("year"),
    names_to = "type",
    values_to = "values")

long_data %>% 
  ggplot(aes(x = year,
             y = values,
             colour = type)) +
  geom_point() +
  geom_path() +
  labs(title = "Can you make this plot using the data above")
```
]

---
## Pivot

<center>
<img src="https://www.fromthebottomoftheheap.net/assets/img/posts/tidyr-longer-wider.gif", width="40%">
</center>

---

# Session 3: modelling with `tidyverse`

---

# Final remarks